python/tvm/relay/frontend/common.py

```diff
@@ -1067,6 +1067,20 @@ def __init__(self, span):
             self._span = tvm.relay.Span(tvm.relay.SourceName(span.decode("utf-8")), 0, 0, 0, 0)
         else:
             assert False, f"unsupported span type: {type(span)}"
+        self.suffix_str = "_PART_"
+        self.counter = 0
+        self.distance_from_leaf = -1
+
+    def _create_span(self):
+        """Adds suffix_str + counter value to _span.source_name.name,
+        to create a unique source_name for the Relay layer
+        """
+        if self.distance_from_leaf == 0:
+            return tvm.relay.Span(tvm.relay.SourceName(self._span), 0, 0, 0, 0)
+        self.distance_from_leaf -= 1
+        span_str = "{}{}{}".format(self._span.source_name.name, self.suffix_str, str(self.counter))
+        self.counter += 1
+        return tvm.relay.Span(tvm.relay.SourceName(span_str), 0, 0, 0, 0)
```

Comment:
> This seems sensible to me, cc the original author of span filling @chunit-quic just to check

Hi all,

Thank you for informing us this change, we just check the PR. We have some ideas about it, and we would like to share some discussions about the suffix in the forum with you.

About the way how span is used in the PR, it looks to me like you would like to take the
source_name of span as an indicator to the original Relay call(expression). 
We find the hash value might be enough for your use case rather than making a unique source_name for a span. Please allow me to describe some problems might be cuased by adding suffix.

We removed suffix in the new implementation in our [preRFC](https://discuss.tvm.apache.org/t/pre-rfc-tvm-explorer-infrastructure/13457#reference-level-explanation-9). In short, we found adding suffix might mislead in some cases. Here is an example:

Suppose:
1. An user thinks span meaning where an expression is from
2. A conv2d layer is converted to 2 relay exprs (conv2d and relu) in a specific frontend.

The conversion of a simple frontend model with only a conv2d layer will result in this:

```python
#FRONTEND
conv2d layer, named as my_conv2d
#Relay IR
%0 = conv2d() /* my_conv2d_part_0 */
%1 = relu() /* my_conv2d_part_1 */
```
In the case above, that user might assume there are "two layers" in the original frontend model, rather than thinking the converted Relay exprs are from the same forntend layer.
One is conv2d layer with the name "my_conv2d_part_0", and a relu layer which is named as
"my_conv2d_part_1".

Yet with help of suffix we do have some advantages. Like allocating the expression which generates the tensor output result of its frontend layer. So far we have [many discussions](https://discuss.tvm.apache.org/t/pre-rfc-tvm-explorer-infrastructure/13457/26?u=chunit)
with community but we are still thinking about whether there is a better way to deal with.

Therefore currently we won't recommend you to add suffix to the source_name of span. After checking the codes, perhaps you could consider using the hash value, `tvm.ir.structural_hash()` of an expr as the key of dictionary, like:

```python
#python/tvm/relay/analysis/operations_distribution.py
class AnalyzeOperationsDistribution(ExprVisitor):
    #...
    def visit_call(self, call: relay.Call):
        if isinstance(call.op, tvm.ir.Op):
            self.unique_op_ids[tvm.ir.structural_hash(call)] = [self.compiler_name, self.func_name, self.func_id]
#python/tvm/driver/tvmc/compiler.py
def dump_operation_offloads(...):
    def annotate_f(x):
        ret = ""
        if isinstance(x, relay.Call):
            compiler_name, op_name, func_id = operations_distribution[tvm.ir.structural_hash(x)]
            ret = f", func_id: {func_id}, compiler_name: {compiler_name}, op_name: {op_name}"
        return ret
```

Because no rewriting/mutation is invoked in `dump_operation_offloads()`, hash value won't be changed. Maybe hash value is an acceptable alternative?

Thank you for reading! Feel free if you have any question. We will try to get back to you soon. :)

Reply:
> Because no rewriting/mutation is invoked in `dump_operation_offloads()`, hash value won't be changed. Maybe hash value is an acceptable alternative?
> 

Hello @chunit-quic, 
I apologize for the delay in replying. Thank you very much for your explanations about suffixes! 
Got it, I'll try to use hash value :)

Reply:
Hello @chunit-quic.

Unfortunately, in my case the value of tvm.ir.structural_hash(call) given at AnalyzeOperationsDistribution:visit_call()
differ from the value of tvm.ir.structural_hash(x) at dump_operation_offloads:annotate_f().

Our task was to print in a human-readable form, which operators from the original graph are offloaded to Ethos-u and which aren't. 
To do this, we collected the information from the Relay after Partition and annotated the original Relay (it is implemented in this PR). And the key point was that we used a suffix to obtain a unique identifier that associates the string of the original Relay with the operator from the resulting Relay. 

But the suffixes will be removed, and the hash does not match.
It looks like we don't have the right mechanism yet to associate the original Relay call with the resulting Relay.

We probably need to wait a decision for this [preRFC](https://discuss.tvm.apache.org/t/pre-rfc-tvm-explorer-infrastructure/13457#reference-level-explanation-9).

Reply:
Hello @lhutton1 , @chunit-quic 

We would propose to accept changes for this PR despite the fact that it is based on the use of suffixes, which in the future will be abolished. The point is that there is also a TVMC option implemented here with the output of which initial Relay operations to which target are offloaded. When in the future the suffix mechanism will be replaced by some other one, we can replace its use for this TVMC option.

Reply:
Apologies for only just getting back to this, yep I think the hash on the call won't work due to difference between `initial_mod` and `mod` caused by mutation during the partitioning process :/ I had a read through the span filling preRFC and it does indeed sound like a complex issue, but I think the reasoning for removing the suffix is valid (see [here](https://discuss.tvm.apache.org/t/pre-rfc-tvm-explorer-infrastructure/13457/29#how-many-benefits-can-we-get-from-suffix-2)), so I'm a bit hesitant to re-introduce it, but would be good to hear @chunit-quic's opinion.

A naive and not well thought through suggestion, but I thought I'd throw it out there - since spans don't seem to be mutated during the lowering, is it possible to make span uniquely hashable, perhaps by introducing some kind of id to the SpanNode and therefore keeping the name intact?

Reply:
Hi @arina-grovety and @lhutton1 

Thank you for informing us at once :D. Sorry that I didn't notice the rewriting part in the invoked pass. Before going to details, our suggestion will be "write a post-processing to tag span suffix" specifically for this kind of scenario. So that we could have a workaround and still keep suffix being discarded as default.
Here are the reasons in short below:
1. We still cannot find methods to handle the impacts after introducing the suffix as default.
2. Span somehow will be changed according to the functionality of pass.

Now I would like to go through the process about why we would suggest writing the post-processing pass. I will try to describe the goals of this PR, and what problems might be encountered. Please kindly correct me if I still miss something. :)

# Goal
Obtain the mapping between the offloaded result and the frontend operators. The flow should be:
```
frontend model ->
Relay IR (IR1) ->
Relay IR transfomred by partition_for_ethosu() (IR2)->
tvmc.complie_model ->
Relay IR transfomred by convert_graph_layout and partition_functions (IR3)->
dump initial_relay offloaded result
```
Based on the codes PR([compiler.py:347](https://github.com/apache/tvm/blob/325b2ef24496eadc14836cb2c8a168a06dff37e8/python/tvm/driver/tvmc/compiler.py#L347)), we would like to get the mapping via IR3 and IR2.

# Requirement
We need a unique indicator to maintain the relation offloaded result. If the `source_name` of `span` is unique, it might be a choice. Therefore here comes two proposals:
1. Add suffix to span as default.
2. Add unique id to span.

# Challenges we will encounter
In this section we would try to describe what challenges we encountered. Since the expressions are modified by partition_functions, we cannot use the hash key of expressions. Now let's check the proposals in the previous section.

1. Add suffix to span as default.  
  The drawback of setting suffix as default is still significant. Like the discussions in the [preRFC](https://discuss.tvm.apache.org/t/pre-rfc-tvm-explorer-infrastructure/13457/29#how-many-benefits-can-we-get-from-suffix-2).
  We still don't encourage to add it back as default.
2. Add unique id to span  
  An obvious obstacle we will face is, it becomes tough to compare the equality between span once we   add an unique id. Another problem is, with the change of expressions, span might be changed also   according to the functionality of pass.
Take the concept of foldconstant pass for example, consider the following transformation:
  ```python
  IR1:
    fn(%input) {
      %0 = add(meta[relay.Constant] /* span=const_1*/, meta[relay.Constant] /* span=const_2*/]) /* span=add*/
      %1 = subtract(%input, %0) /* span=sub */
    }
  # After foldconstant pass it should look like
  IR2:
    fn(%input) {
      %0 = subtract(%input, meta[relay.Constant] /* span=[const_1, const_2, add]*/) /* span=sub */
    }
  ```
  The span of the folded constant expressions is from two Constant and an add Call. Therefore, even   if we have the unique id of const_1, const_2 and add span, we still cannot have a better experience to get the relation.

  Note that if you are interested in how to construct a many-to-one span, it is [one of the ongoing features](https://discuss.tvm.apache.org/t/pre-rfc-tvm-explorer-infrastructure/13457#pass-source-information-builder-6) in the preRFC of TVM Explorer infra. We will try to submit that PR recently.

# Conclusion
From my point of view, it is still challenging for us to add back the suffix, and unique ID
for span might not be easy to deal with either. However, we do observe that making span as an unique indicator is quite useful in some cases. Therefore we will suggest writing a distinct pass to tag suffix to the span. With the help of the suffix-tagging pass, we can have a flexibility to work with and still keep the suffix problem away from the build flow by default.

Here is a sudo persudo code for your reference. 
```python
from collections import defaultdict
class _SuffixTagger(ExprMutator):
    """_SuffixTagger"""
    def __init__(self):
        ExprMutator.__init__(self)
        # key: span or source name, value: counter, indexed from 0
        self.lookup = defaultdict(int)
        self.suffix = "_PART_"
    def _tag_suffix(self, span):
        # To avoid error once we introduce the SequentialSpan in the future
        #https://discuss.tvm.apache.org/t/pre-rfc-tvm-explorer-infrastructure/13457#pass-source-information-builder-6
        # Don't need this if currently
        if isinstance(span, relay.Span):
            ori_name = span.source_name.name
            new_name = ori_name + self.suffix + self.lookup[ori_name]
            lookup[ori_name] += 1
            return tvm.relay.Span(
                tvm.relay.SourceName(new_name), span.line, span.end_line, span.column, span.end_column
            )
        return span
    def visit(self, expr):
        if hasattr(expr, "span"):
            return super().visit(expr)
    def visit_call(self, call):
        new_args = [self.visit(arg) for arg in call.args]
        new_op = self.visit(call.op)
        return _expr.CallWithFields(
            call, new_op, new_args, call.attrs, call.type_args, None, self._tag_suffix()
        )
```

Thank you for reading such long post. If there is any question please don't hesitate to let me know. I will get back to you soon. :D

Reply:
Hello @chunit-quic,
thank you for the detailed explanations!

I'm sorry, but I'm a bit confused. Let me explain my question:

First, do I understand correctly that the workflow in my case will be as follows:

frontend model -> Relay IR (IR1)
if --dump-offloads option is on
     tag IR1 with suffixes using the _SuffixTagger pass
inside the tvmc.compiler.compile_model:
        ...
        IR1(optionally tagged) transfoms by partitions for selected backends (IR2)
        if --dump-offloads option is on
            dump initial_relay offloaded result
        ... 
        
And here is my question:
If I understand correctly, the suffix, although optional, sometimes will still be present in the Relay.
Doesn't this mean that, although optional, we will again get the situation we wanted to avoid by removing the suffixes?

Reply:
Hi @arina-grovety ,

> First, do I understand correctly that the workflow in my case will be as follows:

Yes, invoke the post-processing pass to tag suffix to IR1 should result in the same effect as your modification currently. Which is that the relation between each span and expression is one-to-one mapping.

One more thing is, sorry that perhaps I forgot to deal a condition in the previous pseudo code. We might need a dict to record whether the span of an expression is rewritten. If yes, return the written expression, else keep visiting the expression. Otherwise, we might rewrite the span of an expression which has many consumers mutiple times.

> Doesn't this mean that, although optional, we will again get the situation we wanted to avoid by removing the suffixes?

Unfortunately yes. The post-processing way is a workaround specifically for this case. We will still face the problems introduced by suffix, even though it is optional. Currently we use the span, or to be more precisely, the source_name attribute of span as the one and only one indicator to the source of an expression. Somehow it might be OK for the mapping between `frontend source(TF, ONNX...) <-> Relay IR (with passes)`. 
However, we observe that in some use cases, what we ask for span is much more than what it can provide to us currently. Like:

1. Be an unique indicator for us to obtain which expression a span belongs to.
2. Indicate which expression produces the output tensor of its frontend source layer in one-to-many conversion.
3. Construct precise mapping between a pass

Including the issue 1 we have in here, we also have some discussions in the preRFC try to handle the issues above, for example.
[line column number for 3.](https://discuss.tvm.apache.org/t/pre-rfc-tvm-explorer-infrastructure/13457#unresolved-questions-19)
[A var tracing map for 2](https://discuss.tvm.apache.org/t/pre-rfc-tvm-explorer-infrastructure/13457/17?u=chunit)
[add attribute or custom interface for 2.](https://discuss.tvm.apache.org/t/pre-rfc-tvm-explorer-infrastructure/13457/32#1-extend-the-object-span-add-an-attribute-1)

Because both expressions and spans might be changed, merged or even discarded during the pass transformation, it is hardly to reach all expectations with the capacity of current span instance. Until now we have gathered some idea, yet it is still far from the implementation. Therefore currently a post-processing workaround might be an short-term alternative before a proper mechanism is done